### PR TITLE
Feat: vela adopt support multi-cluster adoption

### LINKED ISSUE #5464: [Feature] vela adopt support multi-cluster adoption
**Is your feature request related to a problem? Please describe.**

By far, vela adopt only support adopt resources on the hub control plane, aka, the local cluster.

In the case users want to take over resources in managed cluster, it is recommended to make some adoption support for that.

**Describe the solution you'd like**

Support `vela adopt deployment/cluster-asia/default/nginx` typed resource adoption. This should automatically retrieve the nginx deployment in cluster-asia's default namespace and add topology policy for that.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by basuotian**
### Description of your changes

Close #5464

I have:

- [x] Read and followed KubeVela's [contribution process](https://github.com/kubevela/kubevela/blob/master/contribute/create-pull-request.md).
- [x] [Related Docs](https://github.com/kubevela/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [ ] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### Special notes for your reviewer

Multi-cluster adoption needs to get resources from sub-cluster, maybe we should have a more elegant implement instead of reusing `multicluster.RequestRawK8sAPIForCluster`:thinking::thinking::thinking:

### CODE REVIEW

File: references/cli/adopt.go [hunk 1]
```diff
@@ -143,8 +146,12 @@ func (opt *AdoptOptions) parseResourceRef(f velacmd.Factory, cmd *cobra.Command,
 	case 3:
 		or.Namespace = parts[1]
 		or.Name = parts[2]
+	case 4:
+		or.Cluster = parts[1]
+		or.Namespace = parts[2]
+		or.Name = parts[3]
 	default:
-		return nil, fmt.Errorf("resource should be like <type>/<name> or <type>/<namespace>/<name>")
+		return nil, fmt.Errorf("resource should be like <type>/<name> or <type>/<namespace>/<name> or <type>/<cluster>/<namespace>/<name>")
```
Comment (chivalryq) [inline references/cli/adopt.go:154]: I think <cluster> should be first argument. The order should be `<cluster>/<type>/<namespace>/<name>`

Comment (basuotian) [reply references/cli/adopt.go:154]: All right, the origin order comes from description in #5464 , i will change the order to `<cluster>/<type>/<namespace>/<name>`

Comment (Somefive) [reply references/cli/adopt.go:154]: I would prefer <type>/<cluster>/<namespace>/<name> as <cluster>/<namespace>/<name> as a whole could be treated as a complete resource identifier.

**Review by Somefive — COMMENTED** [review]
There are some basic directions this PR possible need to change.

1. It would be better not expose `topology` for cue adopt templates. Topology is a higher level concept than cluster and it is not always equal to a single cluster.
2. It would be better to directly use `app.oam.dev/cluster` the annotation on the adopt cluster to pass to adopt templates.
3. I think there will be complex changes for the adopt-templates default.cue. The general idea is to divide resource groups into multiple groups. Each group contain resources from one cluster.

File: references/cli/adopt.go [hunk 2]
```diff
@@ -248,8 +258,19 @@ func (opt *AdoptOptions) loadNative(f velacmd.Factory, cmd *cobra.Command) error
...
 		obj.SetGroupVersionKind(ref.GroupVersionKind)
-		if err := f.Client().Get(cmd.Context(), apitypes.NamespacedName{Namespace: ref.Namespace, Name: ref.Name}, obj); err != nil {
-			return fmt.Errorf("failed to get resource for %s: %w", ref.Arg, err)
+		if ref.Cluster == "" {
+			if err := f.Client().Get(cmd.Context(), apitypes.NamespacedName{Namespace: ref.Namespace, Name: ref.Name}, obj); err != nil {
+				return fmt.Errorf("failed to get resource for %s: %w", ref.Arg, err)
+			}
+		} else {
+			absPath := convertAbsPathFromRef(ref)
+			content, err := multicluster.RequestRawK8sAPIForCluster(cmd.Context(), absPath, ref.Cluster, f.Config())
```
[the remark this answers is not in the document]
Comment (basuotian) [reply references/cli/adopt.go]: all right, change to f.Client().Get(multicluster.WithCluster())

File: references/cli/adopt.go [hunk 3]
```diff
@@ -169,6 +176,9 @@ func (opt *AdoptOptions) Complete(f velacmd.Factory, cmd *cobra.Command, args []
 		if opt.AppNamespace == "" {
 			opt.AppNamespace = opt.NativeResourceRefs[0].Namespace
 		}
+		if opt.NativeResourceRefs[0].Cluster != "" {
+			opt.Topology = opt.NativeResourceRefs[0].Cluster
```
[the remark this answers is not in the document]
Comment (basuotian) [reply references/cli/adopt.go]: rename to `cluster`

File: references/cli/utils.go [hunk 4]
```diff
@@ -240,3 +240,7 @@ func printObj(format string, obj interface{}) (string, error) {
 
 	return ret, nil
 }
+
+func convertAbsPathFromRef(ref *resourceRef) string {
```
[the remark this answers is not in the document]
Comment (basuotian) [reply references/cli/utils.go]: we don't need this function any more as RequestRawK8sAPIForCluster is deprecated...

**Comment by basuotian** [comment]
Reconstruct to support adopt multi-cluster resources in one adoption.
There are two cases with different policy and workflow:

1. adopt multi resources in the same cluster, using topology policy like:
```yaml
  policies:
  - name: read-only
    properties:
      rules:
      - selector:
          componentNames:
          - nginx.Deployment.nginx-sub-cluster
          - nginx.Deployment.nginx-2-sub-cluster
    type: read-only
  - name: topology-sub-cluster
    properties:
      clusters:
      - sub-cluster
    type: topology
```

2. adopt multi resources in different cluster, using apply-component workflowstep like:
```yaml
  policies:
  - name: read-only
    properties:
      rules:
      - selector:
          componentNames:
          - nginx.Deployment.nginx
          - nginx.Deployment.nginx-sub-cluster
    type: read-only
  workflow:
    steps:
    - name: apply-component-nginx.Deployment.nginx
      properties:
        cluster: local
        component: nginx.Deployment.nginx
      type: apply-component
    - name: apply-component-nginx.Deployment.nginx-sub-cluster
      properties:
        cluster: sub-cluster
        component: nginx.Deployment.nginx-sub-cluster
      type: apply-component
```

**Review by wonderflow — COMMENTED** [review]
please add some test cases

**Comment by Somefive** [comment]
@basuotian Let's wait for a bit time because @FogDong is developing another vela adopt capability in parallel now. Generally LGTM, great job!

**Review by Somefive — APPROVED** [review]

**Comment by FogDong** [comment]
Let's hold this until 1.7.6 is released since a new command called `vela adopt --all` will be cherry-picked into 1.7.6.

**Comment by basuotian** [comment]
OK, got it

**Comment by Somefive** [comment]
Let's do some rebase. The PR of `adopt --all` has been merged. @basuotian

**Comment by basuotian** [comment]
> Let's do some rebase. The PR of `adopt --all` has been merged. @basuotian

👌

**Pull request merged**